posta/models.py

```diff
@@ -338,104 +270,63 @@ def invia_raw(cls, oggetto, corpo_html, email_mittente,
-            datetime.now().isoformat(' '),
-            dimensione_massima,
-        ))
-        totale = 0
-        print("%s apro connessione al backed di invio posta" % (
-            datetime.now().isoformat(' '),
-        ))
-        connection = get_connection()
-        for messaggio in da_smaltire:
-            totale += 1
-            if cls.logging:
-                logger.debug('MSG %s: pickup destinatari=(totale=%d, in_attesa=%d)' % (
-                    messaggio.pk, messaggio.oggetti_destinatario.all().count(),
-                    messaggio.oggetti_destinatario.filter(inviato=False).count(),
-                ))
-            print("%s invio messaggio id=%d, destinatari=(totale=%d, in_attesa=%d)" % (
-                datetime.now().isoformat(' '),
-                messaggio.pk,
-                messaggio.oggetti_destinatario.all().count(),
-                messaggio.oggetti_destinatario.filter(inviato=False).count(),
-            ))
-            messaggio.invia(connection)
-        print("%s -- fine elaborazione, elaborati=%d" % (
-            datetime.now().isoformat(' '),
-            totale,
-        ))
-
-
-    @classmethod
-    def costruisci(cls, oggetto='Nessun oggetto', modello='email_vuoto.html', corpo={}, mittente=None, destinatari=[], allegati=[], **kwargs):
-
-        corpo.update({
-            "mittente": mittente,
-            "allegati": allegati,
-        })
-
-        # Accorcia l'oggetto se necessario.
-        if len(oggetto) > cls.LUNGHEZZA_MASSIMA_OGGETTO:
-            oggetto = oggetto[:(cls.LUNGHEZZA_MASSIMA_OGGETTO - len(cls.CARATTERI_RIDUZIONE_OGGETTTO))]
-            oggetto = "%s%s" % (oggetto, cls.CARATTERI_RIDUZIONE_OGGETTTO)
-
-        m = cls(
-            oggetto=oggetto,
-            mittente=mittente,
-            corpo=get_template(modello).render(corpo),
-            **kwargs
-        )
-        m._processa_link()
-        m.save()
-
-        # Se QuerySet, rendi distinti.
-        if isinstance(destinatari, QuerySet):
-            destinatari = destinatari.distinct()
-
-        for d in destinatari:
-            m.aggiungi_destinatario(d)
-
-        for a in allegati:
-            a.oggetto = m
-            a.save()
-
-        return m
-
-    @classmethod
-    def costruisci_e_invia(cls, oggetto='Nessun oggetto', modello='email_vuoto.html', corpo={}, mittente=None, destinatari=[], allegati=[], utenza=False, **kwargs):
+    @staticmethod
+    def smaltisci_coda(dimensione_massima=50):
+        """
+            Metodo usato solo in test, non chiamare direttamente in produzione
+            la coda è gestita da celery
+        :param dimensione_massima: numero di emails da smistare
+        :return: None
+        """
+        from django.core import mail
+        assert hasattr(mail, 'outbox'), 'Questo metodo è da utilzzare solo per unittest'
+        for pk in Messaggio.get_email_batch(dimensione_massima):
+            Messaggio.invia(pk)
```

Comment:
La precedente firma di `invia` conteneva un parametro `connessione` che permetteva la condivisione di una singola connessione SMTP. Ora invece la connessione viene ristabilita ad ogni messaggio in coda. Possiamo ripristinare il comportamento precedente?

C'e' inoltre un motivo specifico per cui `invia` e' diventato un metodo statico che prende `pk`? Qualcosa come,
```python
def get_email_batch(...):
    # ...
    pks = [...]
    return Messaggio.objects.filter(pk__in=pks)
```
avrebbe pochissimo overhead, ma ci permetterebbe di mantenere la firma attuale (e poter chiamare `.invia` su un qualunque oggetto `Messaggio`).

Reply:
con celery non è possibile passare oggetti arbitrari da worker a worker e da django a worker, sono processi diversi, teoricamente possono stare su host diversi, l'unico modo è passare id serializzandoli, una connessione smtp ancora di meno visto che è un socket aperto, l'unico modo per riutilizzare una connessione sarebbe far fare più messaggi per ogni worker (cosa non possibile), non sono a conoscenza della possibilità di salvare un qualcosa nel worker e usarla nelle successive run, forse è possibile ma non so se ne vale la pena

Reply:
@acarmisciano ok, grazie la tua risposta.

Per quanto riguarda la condivisione della connessione SMTP, va benissimo, avevo interpretato male il codice. Credevo per un attimo che il batch venisse smistato in un singolo task, ma vedo che vengono spawnati piu' task `send`, come da nostra richiesta.

Per quanto riguarda invece la firma di `Messaggio.invia` capisco che l'oggetto non puo' essere serializzato, ma allo stesso tempo e' possibile caricare il messagio nel task stesso, e.g.:

```python
for pk in Messaggio.get_email_batch(dimensione_massima):
    msg = Messaggio.objects.get(pk=pk)
    msg.invia()
```

E questo ci permetterebbe di mantenere invariata la firma del metodo `invia`.

Capisco che questo vuol dire che in questo modo viene fatto il fetch due volte del messaggio (una per caricare il messaggio, una per fare il select con locking), ma ne varrebbe comunque la pena - il secondo accesso avrebbe costo insignificante, O(1), perche' gia' in cache, se non comunque al peggio O(log n) dovesse non funzionare la cache, perche' ricerca per chiave indicizzata.

E, volendo, anche questo potrebbe comunque essere risolto *estendendo* la firma di `Messaggio.invia` in modo retrocompatibile, ad esempio con un nuovo kwarg dove si dice al metodo che il locking e' stato gia' fatto, e.g.:

```python
msg = Messaggio.objects.select_for_update(nowait=True) \
                       .get(pk=pk, terminato__isnull=True)
msg.invia(already_locked=True)
```